funlet-call-me/funlet-call-me.js

```diff
@@ -0,0 +1,293 @@
+
+function isDialDone(env, params) {
+  return (typeof params.Dial === "string" );
+}
+exports.input.isDialDone = isDialDone;
+
+// Copied from Forward Funlet
+function getCallStatus(env, params) {
+  const NO_CALL_STATUS = "";
+  return params.DialStatus || params.DialCallStatus || NO_CALL_STATUS;
+}
+exports.input.getCallStatus = getCallStatus;
+
+function getFallbackUrl(env, params) {
+  return params.FailUrl ||
+    env.FUNLET_CALLME_FALLBACK_URL ||
+    MY_FALLBACK_URL;
+}
+exports.input.getFallbackUrl = getFallbackUrl;
+
+// ## Dependencies
+const Twilio = require('twilio');
+
+// ## Output
+exports.output = {};
+
+// Copied from Whisper Funlet
+function spell( numberString ) {
+  const PAUSE = '. ';
+  return numberString.split('').join(PAUSE)+PAUSE;
+}
+exports.output.spell = spell;
+
+/*
+  Function: getWhisperUrl()
+
+  Parameter:
+    params - object, the set of HTTP parameters received by the Funlet
+
+  Returns:
+    string, the URL with parameters to play a message to the recipient of
+    the forwarded call using the Whisper Funlet
+*/
+function getWhisperUrl( params ) {
+  const
+   BASE_WHISPER_URL=".?Whisper=true",
+   SEP="&";
+
+  let whisperUrl = BASE_WHISPER_URL;
+
+  function copyStringParam( name ) {
+    let value = params[name];
+    if ( typeof value === "string" ) {
+      whisperUrl += SEP + name + "=" + encodeURIComponent( value );
+    }
+  }
+
+  copyStringParam( "Message" );
+  copyStringParam( "Language" );
+  copyStringParam( "Voice" );
+
+  return whisperUrl;
+}
+exports.output.getWhisperUrl = getWhisperUrl;
+
+// Copied from Forward Funlet
+function dialForward(response, fallbackUrl, callerId, timeout) {
+  const BASE_URL = ".";
+  let actionUrl = BASE_URL + "?Dial=true";
+  if ( fallbackUrl !== "" ) {
+    actionUrl += "&" + encodeURIComponent(fallbackUrl);
+  }
+  let dialOptions = {
+    action: actionUrl,
+  };
+  if ( callerId !== "" ) {
+    dialOptions.callerId = callerId;
+  }
+  dialOptions.timeout = timeout;
+  return response.dial( dialOptions );
+}
+exports.output.dialForward = dialForward;
+
+/*
+  Function: callMeStage1()
+
+  Parameters:
+    * response - Twilio.twiml.VoiceResponse, Twilio Voice response in progress
+    * forwardingNumber - string, the forwarding number
+    * timeout - number, duration in seconds to let the forwarding call ring
+                before the recipient picks up
+    * whisperUrl - string, action URL to trigger the Whisper Funlet and get
+                   instructions which ask the recipient to accept the call
+    * fallbackUrl - string, URL of a script with further instructions
+                    in case the forwarding call fails
+
+  Response:
+    The response is modified with instructions to forward the call to the
+    given forwarding number, with given timeout, to play a message asking the
+    recipient to accept the call by pressing a key, and to redirect to the
+    given fallback URL if the forwarding call fails.
+*/
+function callMeStage1(
```

Comment:
Can we rename these to something more descriptive?

Reply:
Probably :) These are the functions responsible for a specific stage of the processing. I could change the names to `stage1DoSomething()`, `stage2DoSomethingElse()`, and so on, but it won't be so easy to find something relevant every time.

These are like separate routes that the handler can direct to. Ideally, they would be implemented as separate handlers and deployed to separate paths.

Reply:
I do not have a general solution at this point, so I am keeping these names for now. We can discuss this further in the context of each Funlet, in their own PR.